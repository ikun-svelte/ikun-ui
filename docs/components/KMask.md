---
title: KMask
lang: en-US
---

# KMask

Adds a mask to the specified element.

## Install

::: code-group

```bash [pnpm]
pnpm add @ikun-ui/mask
```

```bash [yarn]
yarn add @ikun-ui/mask
```

```bash [npm]
npm install @ikun-ui/mask
```

:::

## Basic usage

<demo src="../../../../example/mask/basic.svelte" github="Mask"></demo>

## Custom background color

Set the background color of the mask through the color Attribute

<demo src="../../../../example/mask/bg-color.svelte" github="Mask"></demo>

## Mask Props

| Name   | Type                     | Default          | Description                            |
| ------ | ------------------------ | ---------------- | -------------------------------------- |
| color  | `string`                 | `-`              | The background color of the mask       |
| value  | `boolean`                | `false`          | Whether to display the mask            |
| target | `null / HTMLElement`     | `parent element` | The target element covered by the mask |
| cls    | `string`                 | `-`              | Additional class                       |
| attrs  | `Record<string, string>` | `{}`             | Additional attributes                  |

## Mask Slots

| Name    | Description            |
| ------- | ---------------------- |
| default | Customize Mask content |
