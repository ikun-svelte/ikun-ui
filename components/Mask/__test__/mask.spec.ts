import { tick } from 'svelte';
import { afterEach, expect, test, describe, beforeEach } from 'vitest';
import KMask from '../src/index.svelte';

let host: HTMLElement;

const initHost = () => {
	host = document.createElement('div');
	host.setAttribute('id', 'host');
	host.style.width = '100px';
	host.style.height = '100px';
	document.body.appendChild(host);
};
beforeEach(() => {
	initHost();
});
afterEach(() => {
	host.remove();
});

describe('Test: KMask', () => {
	test('props: value', async () => {
		const instance = new KMask({
			target: host
		});
		expect(instance).toBeTruthy();
		expect(host.innerHTML).not.toBeTruthy();

		const instance2 = new KMask({
			target: host,
			props: {
				value: true
			}
		});
		expect(instance2).toBeTruthy();
		expect(host.innerHTML.includes('top: 0px; left: 0px; width: 100%; height: 100%;')).toBeTruthy();
		expect(host.innerHTML).matchSnapshot();
	});

	test('props: target', async () => {
		const target = document.createElement('div');
		target.setAttribute('id', 'target');
		document.body.appendChild(host);
		await tick();
		const instance = new KMask({
			target: host,
			props: {
				target,
				value: true
			}
		});
		await tick();
		expect(instance).toBeTruthy();
		expect(target.style.position).toBe('relative');
		expect(target.style.overflow).matchSnapshot('hidden');
		expect(host.innerHTML).matchSnapshot();
	});

	test('props: color', async () => {
		const instance = new KMask({
			target: host,
			props: {
				value: true,
				color: 'red'
			}
		});
		expect(instance).toBeTruthy();
		expect((host.children[0] as HTMLElement)!.style.backgroundColor).toBe('red');
		expect(host.innerHTML).matchSnapshot();
	});
});
// TODO e2e test
