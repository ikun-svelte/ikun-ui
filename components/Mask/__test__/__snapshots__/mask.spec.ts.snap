// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Test: KMask > props: color 1`] = `"<div style=\\"background-color: red; top: 0px; left: 0px; width: 100%; height: 100%;\\" class=\\"k-mask--base\\"></div>"`;

exports[`Test: KMask > props: target > hidden 1`] = `"hidden"`;

exports[`Test: KMask > props: target 1`] = `"<div class=\\"k-mask--base\\" style=\\"top: 0px; left: 0px; width: 100%; height: 100%;\\"></div>"`;

exports[`Test: KMask > props: value 1`] = `"<div class=\\"k-mask--base\\" style=\\"top: 0px; left: 0px; width: 100%; height: 100%;\\"></div>"`;
